### ISSUE TITLE
One confused point ragarding the "Gated CNN block", what is the "i" in the "Algorithm 1 PyTorch code of Gated CNN block"

### ISSUE THREAD

**Issue by poult-lab** [yuweihao/MambaOut#275]
Thank you to the authors for the impressive work on applying the Gated CNN to visual tasks. This is a very interesting contribution, as demonstrated by the extensive discussions in the “Issues” section.

In the paper, the authors state that “the primary distinction between the Gated CNN [60] and the Mamba block [25] lies in the presence of SSM.” However, in Algorithm 1, I noticed that the variable “i” appears, which I believe should not be included if the goal is to isolate and demonstrate the effectiveness of SSM.

Additionally, Figure 1 seems inconsistent with the algorithm. The figure shows no concatenation operation, yet in Algorithm 1 there is a torch.cat(i, c) step. Could the authors please clarify this discrepancy?

**Comment by yuweihao** [comment]
Hi @poult-lab ,

Thank you so much for your attention. The Conv in Figure 1's Gated CNN can be regarded as a simple type of Inception Depthwise Convolution (InceptionNeXt: https://arxiv.org/abs/2303.16900) with two branches of depthwise conv and identity mapping. InceptionNeXt paper's Table 1 shows that conducting depthconv on partial channels is enough to keep comparable accuracy. Although Inception Depthwise Convolution cannot save FLOPs, it can improve practical speed.

**Comment by poult-lab** [comment]
> Hi [@poult-lab](https://github.com/poult-lab) ,
> 
> Thank you so much for your attention. The Conv in Figure 1's Gated CNN can be regarded as a simple type of Inception Depthwise Convolution (InceptionNeXt: https://arxiv.org/abs/2303.16900) with two branches of depthwise conv and identity mapping. InceptionNeXt paper's Table 1 shows that conducting depthconv on partial channels is enough to keep comparable accuracy. Although Inception Depthwise Convolution cannot save FLOPs, it can improve practical speed.

Now I think I understand — Mambaout [GatedCNN] incorporated some components of the architecture from InceptionNeXt. Thank you very much for your prompt reply. I wish you great success in your future research endeavors.

**Comment by yuweihao** [comment]
In summary, the implemented Gated CNN is depthwise conv on partial channels, ie [DepthWiseConv, IdentityMapping]. It can also be implemented with depthwise convolution on all channels, ie [DepthWiseConv], which will achieve similar accuracy.

**Comment by poult-lab** [comment]
Thank you for the additional explanation. My main point of confusion is this: if we want to prove that SSM is not a good choice for image classification, shouldn’t we simply remove the SSM from GatedCNN—just as the authors stated with “MambaOut”—and then compare whether the results improve or decline?

However, the authors also mention, “It can also be implemented with depthwise convolution on all channels, i.e., [DepthWiseConv], which will achieve similar accuracy.”
Can I understand this as meaning that even without concatenating other channels, using only [DepthWiseConv] (similar to the original Mamba without SSM) can achieve similar accuracy?

**Comment by yuweihao** [comment]
Version 1 [DepthwiseConv(**dim** channels), IdentityMapping((**hidden - dim**) channels)] # the code is shown in https://github.com/yuweihao/MambaOut/blob/main/models/mambaout.py#L119-L152

Version 2 [DepthwiseConv(**hidden** channels)], the code is shown as below


```
class GatedCNNBlock(nn.Module):
    def __init__(self, dim, expansion_ratio=8/3, kernel_size=7, conv_ratio=1.0,
                 norm_layer=partial(nn.LayerNorm,eps=1e-6), 
                 act_layer=nn.GELU,
                 drop_path=0.,
                 **kwargs):
        super().__init__()
        self.norm = norm_layer(dim)
        hidden = int(expansion_ratio * dim)
        self.fc1 = nn.Linear(dim, hidden * 2)
        self.act = act_layer()
        conv_channels = hidden # conduct depthwise conv on all hidden channels
        self.split_indices = (hidden, hidden)
        self.conv = nn.Conv2d(conv_channels, conv_channels, kernel_size=kernel_size, padding=kernel_size//2, groups=conv_channels)
        self.fc2 = nn.Linear(hidden, dim)
        self.drop_path = DropPath(drop_path) if drop_path > 0. else nn.Identity()

    def forward(self, x):
        shortcut = x # [B, H, W, C]
        x = self.norm(x)
        g, c = torch.split(self.fc1(x), self.split_indices, dim=-1)
        c = c.permute(0, 3, 1, 2) # [B, H, W, C] -> [B, C, H, W]
        c = self.conv(c)
        c = c.permute(0, 2, 3, 1) # [B, C, H, W] -> [B, H, W, C]
        x = self.fc2(self.act(g) * c, dim=-1))
        x = self.drop_path(x)
        return x + shortcut
```

Both Version 1 and Version 2 achieve similar accuracy (Version 2 tends to perform marginally better).

**Comment by poult-lab** [comment]
Thank you very very very (very X 100) much for the clear explanation. 
My last question is, why isn’t Version 2 included in the paper if it tends to perform marginally better?

**Comment by yuweihao** [comment]
This is a trade-off. With similar accuracy to Version 2, Version 1 has a higher practical speed (throughput) because it conducts depthwise conv on partial channels.

**Comment by poult-lab** [comment]
Okay now, I understand mamba really outs.
